File: exercise/scala/day08/src/test/scala/doubles/InMemoryToyRepository.scala
package doubles

import domain.{Toy, ToyRepository}

class InMemoryToyRepository extends ToyRepository {
  private val toys = scala.collection.mutable.ListBuffer[Toy]()

  override def findByName(name: String): Option[Toy] = toys.find(_.name == name)

  override def save(toy: Toy): Unit = {
    toys -= toy
    toys += toy
  }
}

File: exercise/php/day08/tests/Doubles/InMemoryToyRepository.php
<?php

namespace Tests\Doubles;

use Domain\Toy;
use Domain\ToyRepository;

class InMemoryToyRepository implements ToyRepository
{
    private array $toys = [];

    public function findByName(string $name): ?Toy
    {
        foreach ($this->toys as $toy) {
            if ($toy->getName() === $name) {
                return $toy;
            }
        }
        return null;
    }

    public function save(Toy $toy): void
    {
        $this->toys = array_filter($this->toys, function ($t) use ($toy) {
            return $t->getName() !== $toy->getName();
        });
        $this->toys[] = $toy;
    }
}
